addons/controllers/antrea/antreaconfig_controller.go

```diff
@@ -0,0 +1,227 @@
+	clusterNamespacedName := req.NamespacedName
+	cluster := &clusterapiv1beta1.Cluster{}
+	for _, owner := range antreaConfig.OwnerReferences {
+		if owner.Kind == cluster.Kind {
+			clusterNamespacedName.Name = owner.Name
+			break
+		}
+	}
+
+	// verify that the cluster related to config is present
+	if err := r.Client.Get(ctx, clusterNamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info(fmt.Sprintf("Cluster '%s'not found in '%s'", clusterNamespacedName.Name, clusterNamespacedName.Namespace))
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, fmt.Sprintf("unable to fetch cluster '%s' in '%s'", clusterNamespacedName.Name, clusterNamespacedName.Namespace))
+		return ctrl.Result{}, err
+	}
+
+	if retResult, err := r.ReconcileAntreaConfig(ctx, antreaConfig, cluster, log); err != nil {
+		log.Error(err, "unable to reconcile AntreaConfig")
+		return retResult, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *AntreaConfigReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&cniv1alpha1.AntreaConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+// ReconcileAntreaConfig reconciles AntreaConfig CR
+func (r *AntreaConfigReconciler) ReconcileAntreaConfig(
+	ctx context.Context,
+	antreaConfig *cniv1alpha1.AntreaConfig,
+	cluster *clusterapiv1beta1.Cluster,
+	log logr.Logger) (_ ctrl.Result, retErr error) {
+
+	var patchConfig bool
+
+	patchHelper, err := clusterapipatchutil.NewHelper(antreaConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Patch AntreaConfig before returning the function
+	defer func() {
+		// patchConfig will be true if finalizer or owner reference is added or deleted
+		if patchConfig {
+			log.Info("Patching AntreaConfig")
+
+			if err := patchHelper.Patch(ctx, antreaConfig); err != nil {
+				log.Error(err, "Error patching AntreaConfig")
+				retErr = err
+			}
+		}
+	}()
+
+	// If AntreaConfig is marked for deletion, then no reconciliation is needed
+	if !antreaConfig.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ReconcileAntreaConfigNormal(ctx, antreaConfig, cluster, log, &patchConfig); err != nil {
+		log.Error(err, "Error reconciling AntreaConfig to create data value secret")
+		return ctrl.Result{}, err
+	}
+
+	log.Info("Successfully reconciled AntreaConfig")
+	return ctrl.Result{}, nil
+}
+
+// ReconcileAntreaConfigNormal reconciles AntreaConfig by creating/patching data values secret
+func (r *AntreaConfigReconciler) ReconcileAntreaConfigNormal(
+	ctx context.Context,
+	antreaConfig *cniv1alpha1.AntreaConfig,
+	cluster *clusterapiv1beta1.Cluster,
+	log logr.Logger,
+	patchConfig *bool) (retErr error) {
+
+	// Add finalizer to addon secret
+	*patchConfig = util.AddFinalizerToCR(log, constants.AntreaAddonName, antreaConfig)
```

Comment:
Do we need a finalizer if deletion is no-op for the controller?

Reply:
This is not needed.nice catch.